src/sentry/api/endpoints/project_dynamic_sampling.py

```diff
@@ -0,0 +1,194 @@
+
+    def get(self, request: Request, project) -> Response:
+        """
+        Generates distribution function values for client sample rates from a random sample of
+        root transactions, and provides the projects breakdown for these transaction when
+        creating a dynamic sampling rule for distributed traces.
+        ``````````````````````````````````````````````````
+
+        :pparam string organization_slug: the slug of the organization the
+                                          release belongs to.
+        :qparam string query: If set, this parameter is used to filter root transactions.
+        :qparam string sampleSize: If set, specifies the sample size of random root transactions.
+        :qparam string distributedTrace: Set to distinguish the dynamic sampling creation rule
+                                    whether it is for distributed trace or single transactions.
+        :qparam string statsPeriod: an optional stat period (can be one of
+                                    ``"24h"``, ``"14d"``, and ``""``).
+        :auth: required
+        """
+        if not features.has(
+            "organizations:filters-and-sampling", project.organization, actor=request.user
+        ):
+            return Response(
+                {
+                    "detail": [
+                        "Dynamic sampling feature flag needs to be enabled before you can perform "
+                        "this action."
+                    ]
+                },
+                status=404,
+            )
+
+        query = request.GET.get("query", "")
+        requested_sample_size = min(int(request.GET.get("sampleSize", 100)), 1000)
+        distributed_trace = request.GET.get("distributedTrace", "1") == "1"
+        stats_period = min(
+            parse_stats_period(request.GET.get("statsPeriod", "1h")), timedelta(hours=24)
+        )
+
+        end_time = timezone.now()
+        start_time = end_time - stats_period
+
+        # Fetches a random sample of root transactions of size `sample_size` in the last period
+        # defined by `stats_period`. The random sample is fetched by ordering by the
+        # `random_number` which is generated by a `random_number` function which generates a random
+        # number for every row. The goal here is to fetch the transaction ids, their sample rates
+        # and their trace ids.
+        root_transactions = discover.query(
+            selected_columns=[
+                "id",
+                "trace",
+                "trace.client_sample_rate",
+                "random_number() AS random_number",
+            ],
+            query=f"{query} event.type:transaction !has:trace.parent_span_id",
+            params={
+                "start": start_time,
+                "end": end_time,
+                "project_id": [project.id],
+                "organization_id": project.organization,
+            },
+            orderby=["-random_number"],
```

Comment:
I fear that `orderby rand() limit 100` will still load the entire time duration into memory in order to sort it. I think there's a chance this query will run out of memory for certain projects, including our own. We'lll have to see because I don't know of a better soln right now

Reply:
I believe the correct solution is to add a `SAMPLE` clause for large projects. I just do not know how to estimate the right number here.

Reply:
What if I apply a sample that is 10x the query limit? must be better than not applying any sample

Reply:
that sounds reasonable, you'd have to try whether the results work out for you

Reply:
@ahmedetefy i think SAMPLE samples across the entire data set. So if you only look at 48h but you SAMPLE to 10000 then you getting ~10000 samples across the entire 90 days before the filter applies.

Reply:
I think if you find a way to do sort of (stateful?) sampling in the `WHERE` clause you should be good. `WHERE rand() < 0.5` doesn't have those memory problems, but you don't want 50% of values, you want `n` values

Reply:
oh. you could first run `select count() ... where <timerange>` to determine how many rows you're scanning, then calculate a sample rate that you can pass to this query. this would get rid of potential OOM

Reply:
Do we have any idea what the threshold is before OOM?

Reply:
no, and you shouldn't rely on that